### PR TITLE
MGMT-3108: Adding bootstrap-in-place installation support

### PR DESCRIPTION
**Pull request by osherdp**

### CODE REVIEW

File: discovery-infra/resources/install-config.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+apiVersion: v1
```
Comment (romfreiman) [inline discovery-infra/resources/install-config.yaml:1]: ipv6 will be required pretty soon

File: discovery-infra/resources/sno_manifest.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+apiVersion: machineconfiguration.openshift.io/v1
```
Comment (romfreiman) [inline discovery-infra/resources/sno_manifest.yaml:1]: I think it's not required anymore. @eranco74 ?

File: discovery-infra/test_infra/utils.py [hunk 3]
```diff
@@ -306,14 +309,14 @@ def wait_till_at_least_one_host_is_in_status(
...
 def wait_till_specific_host_is_in_status(
-    client,
-    cluster_id,
-    host_name,
-    nodes_count,
-    statuses,
-    timeout=consts.NODES_REGISTERED_TIMEOUT,
-    fall_on_error_status=True,
-    interval=5,
+        client,
```
Comment (romfreiman) [inline discovery-infra/test_infra/utils.py:312]: now u like @tsorya  :)

File: discovery-infra/resources/install-config.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  replicas: 1
+metadata:
+  creationTimestamp: null
+  name: test-infra-cluster
+networking:
+  clusterNetwork:
+  - cidr: 10.128.0.0/14
+    hostPrefix: 23
+  machineNetwork:
+  - cidr: 192.168.126.0/24
```
Comment (tsorya) [inline discovery-infra/resources/install-config.yaml:23]: we need to change ip here as in terraform

File: Makefile [hunk 5]
```diff
@@ -283,6 +283,11 @@ destroy_all_nodes_from_namespaces:
 destroy_all_nodes:
 	skipper run $(SKIPPER_PARAMS) 'discovery-infra/delete_nodes.py --delete-all'
 
+deploy_ibip: _test_setup
+	skipper make $(SKIPPER_PARAMS) _deploy_nodes $(SKIPPER_PARAMS) ADDITIONAL_PARAMS="'--bootstrap-in-place'" NUM_WORKERS=0 NUM_MASTERS=1 NAMESPACE_INDEX=0
+
+wait_for_ibip_installation_completion:
```
Comment (tsorya) [inline Makefile:289]: why not to make it as part of deploy ibip ?

Comment (tsorya) [reply Makefile:289]: @RazRegev PR is adding platform field i think we just need to move to it and not to use new flag as --bootstrap-in-place.
What do u think? https://github.com/openshift/assisted-test-infra/pull/356/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R97

File: terraform_files/main.tf [hunk 6]
```diff
@@ -132,3 +140,27 @@ resource "libvirt_domain" "worker" {
     dev = ["hd", "cdrom"]
   }
 }
+
+data "libvirt_network_dns_host_template" "api" {
```
Comment (tsorya) [inline terraform_files/main.tf:144]: lets add nice comment for future use , how it works. I think we will have more stuff like this and be great that we will have explanation for this beautiful code

File: Makefile (hunk 5 above)
Comment (osherdp) [reply Makefile:289]: It's not a platform, BIP is an installation mode
And I also think that we shouldn't make it similar to the regular flow, not until it's integrated in AI

Comment (osherdp) [reply Makefile:289]: As for integrating ``wait_for_ibip_installation_completion`` as a part of ``deploy_ibip``, done

File: discovery-infra/resources/install-config.yaml (hunk 1 above)
Comment (osherdp) [reply discovery-infra/resources/install-config.yaml:1]: ok

File: discovery-infra/resources/install-config.yaml (hunk 4 above)
Comment (osherdp) [reply discovery-infra/resources/install-config.yaml:23]: fixed

File: terraform_files/main.tf (hunk 6 above)
Comment (osherdp) [reply terraform_files/main.tf:148]: done

File: discovery-infra/bootstrap_in_place.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+        logging.info("Waiting for installation to complete...")
+        waiting.wait(all_operators_up,
+                     sleep_seconds=20,
+                     timeout_seconds=60 * 60,
+                     waiting_for="all operators to get up")
+        logging.info("Installation completed successfully!")
+
+    finally:
+        if host_ip is not None:
```
Comment (tsorya) [inline discovery-infra/bootstrap_in_place.py:165]: can we push all this to function? and just call it?

File: discovery-infra/bootstrap_in_place.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+        download_live_image(f"{BUILD_DIR}/installer-image.iso")
+        image_path = embed("installer-image.iso", "bootstrap.ign", EMBED_IMAGE_NAME)
+
+        logging.info("Starting nodes...")
+        controller.image_path = image_path
+        controller.start_all_nodes()
+        logging.info("Nodes started!")
+
+        logging.info("Configuring /etc/hosts...")
```
Comment (tsorya) [inline discovery-infra/bootstrap_in_place.py:151]: can we have waiting for installation to complete function?

File: discovery-infra/bootstrap_in_place.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    all_operators_are_valid = len(invalid_operators) == 0
+
+    if not all_operators_are_valid:
+        logging.debug("Following operators are still down: %s", ", ".join(invalid_operators))
+
+    return all_operators_are_valid
+
+
+def gather_sosreport_data(host_ip):
+    with SshConnection(ip=host_ip, private_ssh_key_path=SSH_KEY) as ssh:
```
Comment (tsorya) [inline discovery-infra/bootstrap_in_place.py:122]: we have this option in Node object. Why not to use the object?

Comment (tsorya) [reply discovery-infra/bootstrap_in_place.py:122]: and add function that will upload file will be very helpful to there

File: discovery-infra/bootstrap_in_place.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        ssh.upload_file(SOSREPORT_SCRIPT, "/tmp/man_sosreport.sh")
+        ssh.execute("chmod a+x /tmp/man_sosreport.sh")
+        ssh.execute("sudo /tmp/man_sosreport.sh")
+        ssh.download_file("/tmp/sosreport.tar.bz2", IBIP_DIR)
```
Comment (tsorya) [inline discovery-infra/bootstrap_in_place.py:126]: same here

File: discovery-infra/bootstrap_in_place.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+SOSREPORT_SCRIPT = os.path.join(RESOURCES_DIR, "man_sosreport.sh")
+SSH_KEY = os.path.join("ssh_key", "key")
+
+
+def installer_generate():
+    logging.info("Installer generate manifests")
+    utils.run_command(f"{INSTALLER_BINARY} create manifests --dir={IBIP_DIR}")
+    logging.info("Installer generate ignitions")
+    # TODO delete
+    shutil.copy(f"{RESOURCES_DIR}/sno_manifest.yaml", os.path.join(IBIP_DIR, "openshift"))
```
Comment (tsorya) [inline discovery-infra/bootstrap_in_place.py:34]: @eranco74 still needed?

File: discovery-infra/resources/install-config.yaml (hunk 1 above)
Comment (osherdp) [reply discovery-infra/resources/install-config.yaml:1]: we will handle it in the future

File: discovery-infra/resources/sno_manifest.yaml (hunk 2 above)
Comment (osherdp) [reply discovery-infra/resources/sno_manifest.yaml:1]: @eranco74 will soon change it with the new payload

File: discovery-infra/bootstrap_in_place.py (hunk 9 above)
Comment (osherdp) [reply discovery-infra/bootstrap_in_place.py:122]: ok

File: discovery-infra/bootstrap_in_place.py (hunk 8 above)
Comment (osherdp) [reply discovery-infra/bootstrap_in_place.py:151]: yes

File: discovery-infra/bootstrap_in_place.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    logging.info("Starting node...")
+    controller.image_path = image_path
+    controller.start_all_nodes()
+    node = Nodes(controller, private_ssh_key_path=SSH_KEY)[0]
+    logging.info("Node started!")
+
+    try:
+        waiting_for_installation_completion(controller)
+    finally:
+        logging.info("Gathering sosreport data from host...")
```
Comment (tsorya) [inline discovery-infra/bootstrap_in_place.py:167]: can we add this finally code to function too?

Comment (osherdp) [reply discovery-infra/bootstrap_in_place.py]: done

File: discovery-infra/bootstrap_in_place.py (hunk 11 above)
Comment (eranco74) [reply discovery-infra/bootstrap_in_place.py:35]: I'll remove it once we move to quay.io/eranco74/ocp-release:bootstrap-in-place

File: discovery-infra/bootstrap_in_place.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+        logging.info("Waiting for installation to complete...")
+        waiting.wait(all_operators_up,
+                    sleep_seconds=20,
+                    timeout_seconds=60 * 60,
+                    waiting_for="all operators to get up")
+        logging.info("Installation completed successfully!")
+
+    finally:
+        logging.info("Gathering sosreport data from host...")
+        node = Nodes(controller, private_ssh_key_path=SSH_KEY)[0]
```
Comment (eranco74) [inline discovery-infra/bootstrap_in_place.py:145]: Can we log the `oc describe co` here?

**Review by eranco74 — APPROVED** [review]

File: discovery-infra/bootstrap_in_place.py (hunk 13 above)
Comment (osherdp) [reply discovery-infra/bootstrap_in_place.py:145]: It's in the sosreport script

**Pull request merged**